clang/include/clang/Sema/Sema.h

```diff
@@ -1759,6 +1759,33 @@ class Sema final {
     }
   };
 
+  /// Bitmask to contain the list of reasons a single diagnostic should be
+  /// emitted, based on its language.  This permits multiple offload systems
+  /// to coexist in the same translation unit.
+  enum DeviceDiagnosticReason {
+    /// Diagnostic doesn't apply to anything. Included for completeness, but
+    /// should make this a no-op.
+    DDR_None = 0,
+    /// OpenMP specific diagnostic.
+    DDR_OMP_DEVICE = 1 << 0,
+    DDR_OMP_HOST = 1 << 1,
+    DDR_OMP_ALL = DDR_OMP_DEVICE | DDR_OMP_HOST,
+    /// CUDA specific diagnostics.
+    DDR_CUDA_DEVICE = 1 << 2,
+    DDR_CUDA_HOST = 1 << 3,
+    DDR_CUDA_ALL = DDR_CUDA_DEVICE | DDR_CUDA_HOST,
+    /// SYCL specific diagnostic.
+    DDR_SYCL = 1 << 4,
+    /// ESIMD specific diagnostic.
+    DDR_ESIMD = 1 << 5,
```

Comment:
why SYCL and ESIMD are different from OMP and CUDA - no device/host split?

Reply:
I hadn't noticed any diagnostics that seemed to care about host vs device in this case.  Or at least, there aren't really any 'host specific' diagnostics I could find.

Reply:
> I hadn't noticed any diagnostics that seemed to care about host vs device in this case. Or at least, there aren't really any 'host specific' diagnostics I could find.

I see. I assume nothing will prevent us adding those in future when needed.